airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/components.py

```diff
@@ -0,0 +1,103 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+import datetime
+from dataclasses import dataclass
+from typing import List, Optional, Union
+
+import dpath
+from airbyte_cdk.sources.declarative.incremental import DatetimeBasedCursor
+from airbyte_cdk.sources.declarative.transformations import AddFields
+from airbyte_cdk.sources.declarative.types import StreamSlice, StreamState
+from airbyte_cdk.sources.types import Config, Record, StreamSlice, StreamState
+from isodate import Duration, parse_duration
+
+
+@dataclass
+class SnapchatMarketingDatetimeBasedCursor(DatetimeBasedCursor):
+    """
+    SnapchatMarketingDatetimeBasedCursor is a variant of DatetimeBasedCursor that strictly compares start and end dates.
+    It ensures that the start date is always less than the end date to prevent errors.
+
+    Inherits from:
+        DatetimeBasedCursor
+
+    Custom Methods:
+        _partition_daterange: Return slices without intersections
+    """
+
+    def _partition_daterange(self, start: datetime, end: datetime, step: Union[datetime.timedelta, Duration]) -> List[StreamSlice]:
+        start_field = self._partition_field_start.eval(self.config)
+        end_field = self._partition_field_end.eval(self.config)
+        dates = []
+
+        while start < end:
+            next_start = self._evaluate_next_start_date_safely(start, step)
+            end_date = self._get_date(next_start - self._cursor_granularity, end, min)
+
+            dates.append(
+                StreamSlice(
+                    partition={}, cursor_slice={start_field: self._format_datetime(start), end_field: self._format_datetime(end_date)}
+                )
+            )
+
+            start = next_start
+
+        return dates
+
+
+@dataclass
+class AddExistingRecordFields(AddFields):
+    """
+    AddExistingRecordFields is specifically tailored for analytics data, addressing the limitations of the current
```

Comment:
what would the behavior be if we tried to add a field that doesn't exist? wouldn't it just result in a None field? if that's the case, I don't think a custom component is worth the effort as the resulting record should be the same in the destination.

Destinations would actually prefer we leave none fields, so if excluding fields with none values is the only benefit of the custom component, let's just use the default. [link to slack thread for context](https://airbytehq-team.slack.com/archives/C06FQEBNMEY/p1712778408904579)

Reply:
The outcome depends on how we handle cases where a field does not exist in the record. If we don't handle it, the field will be added with an empty string (e.g., ''). Alternatively, if we set a default value (e.g., None). So we use custom component to avoid having many empty or None fields.

Reply:
To clarify, do we need to return a record with the maximum number of fields as described in the schema so that the destination can properly parse the types?

Reply:
ok it seems like we might want a transform specific to moving or copying an existing field from the record. can you create a follow up issue for this one since it seems a little involved?

Reply:
> To clarify, do we need to return a record with the maximum number of fields as described in the schema so that the destination can properly parse the types?
Destinations _should_ handle missing fields correctly today, but it's not a guarantee from the protocol, hence the long-term concern.

Reply:
I replaced the dict indexing to `get` method and now non-existent fields process as None